### PR TITLE
Tls13 write end of early data

### PR DESCRIPTION
**Pull request by xkqian**
1. If the client parse the "early_data" extension in EncryptedExtensions,
   it MUST send an EndOfEarlyData message after receiving the
   server Finished.
2. If the client does not get an "early_data"
   extension in EncryptedExtensions, then it MUST NOT send an
   EndOfEarlyData message

TODO: With current test case, we still can't make the handshake successfully cause gnutls report one error which receive the unexpected packet when processing client finish.  
Anyway, with current test case, we can see the server side can parse and receive the end_of_early_data correctly.

### CODE REVIEW

**Comment by yuhaoth** [comment]
Please preceding #6538 . That message depends on early secrets . 

Before #6623 is resolved and #6538 is merged, I can not approve it.  There is no test cases can guarantee the code is correct

**Review by yuhaoth — CHANGES_REQUESTED** [review]
Please preceding #6538 . This message depends on early secrets . 

Before #6623 is resolved and #6538 is merged, I can not approve it.  There is no test cases can guarantee the code is correct.

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
I think apart from the intermittent issue #6623, with this PR the ssl-opt.sh "TLS 1.3 m->G: EarlyData: basic check, good" test case should run successfully (handshake completed). I have tried it but the GnuTLS server does not seem to be happy with what it receives. Please have a look.

**Comment by yuhaoth** [comment]
Please merge this PR after #6787, that is workaround for #6623

**Comment by xkqian** [comment]
> I think apart from the intermittent issue #6623, with this PR the ssl-opt.sh "TLS 1.3 m->G: EarlyData: basic check, good" test case should run successfully (handshake completed). I have tried it but the GnuTLS server does not seem to be happy with what it receives. Please have a look.

-- With current test case, we still can't make the handshake successfully cause gnutls report one error which receive the unexpected packet when processing client finish.
Anyway, with current test case, we can see the server side can parse and receive the end_of_early_data correctly.

**Comment by yuhaoth** [comment]
> -- With current test case, we still can't make the handshake successfully cause gnutls report one error which receive the unexpected packet when processing client finish. Anyway, with current test case, we can see the server side can parse and receive the end_of_early_data correctly.

I agree with @ronald-cron-arm . `basic check` should pass . And the error  means we have some fatal error.

File: library/ssl_tls13_client.c
Comment (yuhaoth) [inline library/ssl_tls13_client.c:2211]:
```suggestion
```
I do not think we need `coordinate` in EOED

File: library/ssl_tls13_client.c
Comment (yuhaoth) [inline library/ssl_tls13_client.c:2164]:
```suggestion
/*
 * Handler for MBEDTLS_SSL_END_OF_EARLY_DATA
 *
 * RFC 8446 section 4.5 
 * 
 * struct {} EndOfEarlyData; 
 * 
 * If the server sent an "early_data" extension in EncryptedExtensions, the 
 * client MUST send an EndOfEarlyData message after receiving the server 
 * Finished.
 */
```

**Review by yuhaoth — CHANGES_REQUESTED** [review]
Just a quick look. some functions should be re-considered. 
And this PR equals without 0-RTT and with early data indication case, it can complete handshake successfully.  Please check basic test.

File: library/ssl_tls.c
Comment (xkqian) [reply library/ssl_tls.c:1676]: > it should not be set here.
> `chiphersuite_info` is decided in ClientHello and `key_exchange_mode` is decided in `ServerHello` .
In client hello, we will send a list of cipher suite, so it can't be decided in ClientHello, instead, we decide it in parse_ServerHello. 
Right, `key_exchange_mode` is decided in `ServerHello`.  But as we all know, we should compute_early_transform after client hello, which is before parse_ServerHello, unfortunately, compute_early_transform depend on the key_exchange_mode indirectly.  So we have to set the key_exchange_mode before.
Set_session happened when re-connection, we can set the info here.

File: library/ssl_tls13_generic.c
Comment (xkqian) [reply library/ssl_tls13_generic.c:1624]: All of the client_hello writing have been moved to ssl_client.c.  We should call  **ssl_tls13_ticket_get_psk** or **mbedtls_ssl_tls13_ticket_get_psk** there. Any suggestion how we only need in `ssl_tls13_client.c`

File: library/ssl_tls13_client.c
Comment (xkqian) [reply library/ssl_tls13_client.c:2510]: In current design, we just always goto MBEDTLS_SSL_END_OF_EARLY_DATA, and we can decide to skip it or not depend on the earlydata status.

**Comment by xkqian** [comment]
> I agree with @ronald-cron-arm . `basic check` should pass . And the error means we have some fatal error.

Agree. We should make the `basic check` pass. 
Currently, not sure why gnutls server report "Error in handshake: An unexpected TLS packet was received."  I should further investigate the gnutls logic and fix it. Maybe I can write some early data to it to see whether some differences.

**Comment by xkqian** [comment]
The 'basic good' test case of earlydata can be executed successfully locally.

File: library/ssl_tls.c
Comment (yuhaoth) [reply library/ssl_tls.c:1676]: > > it should not be set here.
> > `chiphersuite_info` is decided in ClientHello and `key_exchange_mode` is decided in `ServerHello` .
> 
> In client hello, we will send a list of cipher suite, so it can't be decided in ClientHello, instead, we decide it in parse_ServerHello. Right, `key_exchange_mode` is decided in `ServerHello`. But as we all know, we should compute_early_transform after client hello, which is before parse_ServerHello, unfortunately, compute_early_transform depend on the key_exchange_mode indirectly. So we have to set the key_exchange_mode before. Set_session happened when re-connection, we can set the info here.

move to pre_shared_key not here

File: library/ssl_client.c
Comment (xkqian) [reply library/ssl_client.c:1014]: Do you mean we can decide the **ciphersuit_info** when `write client hello`?  Seems it's not the right logic. we should get confirmed from server side and then select the right **ciphersuite**.
Per my understanding, the ciphersuite should be the ciphersuite saved by last session, and then is loaded when resumption, and then is setted in set_session.   Is that right?

File: library/ssl_tls13_client.c
Comment (xkqian) [reply library/ssl_tls13_client.c]: No need of "K_send = handshake" here. Update done.

**Comment by xkqian** [comment]
When build_armcc report error:"Error: Error: C9555E: C9555E: Failed to check out a license." Seems this failure is caused by upstream, nothing to do  with this PR.

File: library/ssl_client.c
Comment (ronald-cron-arm) [inline library/ssl_client.c:991]: If offering early data and if the middlebox compatibility mode is enabled, we need to send a CCS immediately after the first ClientHello, probably through a MBEDTLS_SSL_CLIENT_CCS_AFTER_CLIENT_HELLO state like in the prototype.

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]

File: library/ssl_tls13_client.c
Comment (yuhaoth) [reply library/ssl_tls13_client.c]: `MBEDTLS_SSL_CLIENT_CERTIFICATE` is not always available. 

For time being, it is available. But there is a coordinate operation, that can be optimized in future.

File: library/ssl_client.c
Comment (yuhaoth) [reply library/ssl_client.c]: It should be only done when `early_data_status = REJECTED`

File: library/ssl_tls13_client.c
Comment (yuhaoth) [inline library/ssl_tls13_client.c:995]:
```suggestion
```

It is not needed

File: library/ssl_tls13_client.c
Comment (yuhaoth) [inline library/ssl_tls13_client.c:993]:
```suggestion
       if( ssl->early_data_satus == MBEDTLS_SSL_EARLY_DATA_STATUS_REJECTED )
        {
            MBEDTLS_SSL_DEBUG_MSG(
                1, ( "Set hs psk for early data when writing the first psk" ) );

            ret = ssl_tls13_ticket_get_psk( ssl, &hash_alg, &psk, &psk_len );
            if( ret != 0 )
            {
                MBEDTLS_SSL_DEBUG_RET(
                        1, "ssl_tls13_ticket_get_psk", ret );
                return ( ret );
            }

            ret = mbedtls_ssl_set_hs_psk( ssl, psk, psk_len );
            if( ret  != 0 )
            {
                MBEDTLS_SSL_DEBUG_RET( 1, "mbedtls_ssl_set_hs_psk", ret );
                return ( ret );
            }

            ciphersuite_info = mbedtls_ssl_ciphersuite_from_id(
                                    ssl->session_negotiate->ciphersuite );
            ssl->handshake->ciphersuite_info = ciphersuite_info;
        }
```

File: library/ssl_tls13_client.c
Comment (yuhaoth) [inline library/ssl_tls13_client.c:2504]: move `K_send = handshake` here

File: library/ssl_tls13_client.c
Comment (yuhaoth) [inline library/ssl_tls13_client.c:2195]: Add `K_send = handshake` here.

File: include/mbedtls/ssl.h
Comment (yuhaoth) [inline include/mbedtls/ssl.h:536]:
```suggestion
#define MBEDTLS_SSL_HS_END_OF_EARLY_DATA        5 
```
Please remove the comment in 537 also.

I prefer remove the comment in `development` or add comment for other handshake message also.

**Review by yuhaoth — CHANGES_REQUESTED** [review]
Some improvements

File: include/mbedtls/ssl.h
Comment (xkqian) [reply include/mbedtls/ssl.h:536]: > Please remove the comment in 537 also.
> 
> I prefer remove the comment in `development` or add comment for other handshake message also.

Can you elaborate which of other handshake messages need to add comment? 
List them here:
```
#define MBEDTLS_SSL_HS_HELLO_REQUEST            0
#define MBEDTLS_SSL_HS_CLIENT_HELLO             1
#define MBEDTLS_SSL_HS_SERVER_HELLO             2
#define MBEDTLS_SSL_HS_HELLO_VERIFY_REQUEST     3
#define MBEDTLS_SSL_HS_NEW_SESSION_TICKET       4
#define MBEDTLS_SSL_HS_END_OF_EARLY_DATA        5 // NEW IN TLS 1.3
#define MBEDTLS_SSL_HS_ENCRYPTED_EXTENSIONS     8 // NEW IN TLS 1.3
#define MBEDTLS_SSL_HS_CERTIFICATE             11
#define MBEDTLS_SSL_HS_SERVER_KEY_EXCHANGE     12
#define MBEDTLS_SSL_HS_CERTIFICATE_REQUEST     13
#define MBEDTLS_SSL_HS_SERVER_HELLO_DONE       14
#define MBEDTLS_SSL_HS_CERTIFICATE_VERIFY      15
#define MBEDTLS_SSL_HS_CLIENT_KEY_EXCHANGE     16
#define MBEDTLS_SSL_HS_FINISHED                20
#define MBEDTLS_SSL_HS_MESSAGE_HASH           254
```

File: library/ssl_tls13_client.c
Comment (xkqian) [reply library/ssl_tls13_client.c:995]: We need this mode be set, cause we need the psk enabled check in early data processing, which is earlier than the process_server_hello().

Comment (yuhaoth) [reply library/ssl_tls13_client.c:995]: That's `conf->kex_mode` not `handshake->key_exchange_mode` .  

`handshake->key_exchange_mode` was determined by negotiate. The value is only valid after server hello. 

And what's  breaked without it?

File: include/mbedtls/ssl.h
Comment (yuhaoth) [reply include/mbedtls/ssl.h:536]: All handshake messages , mark where it is come from. like RFC 8446 , RFC 5246 and so on.  The quickest way is remove the comment.

**Comment by yuhaoth** [comment]
This PR should precedding #6788 .  `TLS 1.3 m->G: EarlyData: basic check, good` is skipped. No test available without #6788

File: library/ssl_tls13_client.c
Comment (xkqian) [reply library/ssl_tls13_client.c:995]: > That's `conf->kex_mode` not `handshake->key_exchange_mode` .
> 
---------No, it need handshake->key_exchange_mode, not only conf->kex_mode.
> `handshake->key_exchange_mode` was determined by negotiate. The value is only valid after server hello.
> 
> And what's breaked without it?
-------The basic good test case, client side will report error: "! bad client exit code (expected 0, got 1)", server side report: "Error in handshake: Decryption has failed."

Comment (yuhaoth) [reply library/ssl_tls13_client.c]: So, add comments to describe the issue.

**Review by yuhaoth — CHANGES_REQUESTED** [review]
For the ciphersuite check part, I think we can compare `ciphersuite_info->id` with `session_negotiate->ciphersuite` to make sure it is same.

For `selected_identity == 0` , it should be move the `postprocess_server_hello`

File: library/ssl_client.c
Comment (xkqian) [reply library/ssl_client.c:1008]: Cause we are not very confidence about the moving, how about leave it here for temp?

From the original logic, seems except the case of tls13_only, we should set the state to MBEDTLS_SSL_SERVER_HELLO  after client hello.

Comment (yuhaoth) [reply library/ssl_client.c:1008]: The origin code has some fatal error. 

- `if( ( ssl->handshake->min_tls_version == MBEDTLS_SSL_VERSION_TLS1_3 ) && ( ssl->tls_version == MBEDTLS_SSL_VERSION_TLS1_3 ) )` can be replace with `mbedtls_ssl_is_tls13_only`
- If `PROTO_TLS12` is enabled and configured as tls12_only, `mbedtls_ssl_tls13_finalize_write_client_hello` will be called. That's wrong
- In hybrid mode , CCS should be sent.

**Review by yuhaoth — CHANGES_REQUESTED** [review]
https://github.com/Mbed-TLS/mbedtls/pull/6499/files#r1066799622 , It MUST be fixed. 
And no tests for ciphersuite mismatch? If possible, please add it.

Other part looks okay for me

File: library/ssl_tls13_client.c
Comment (ronald-cron-arm) [inline library/ssl_tls13_client.c:2583]: We should not go into that state if we have proposed early data (but were rejected) as encryption is in place for outgoing data and we would send an encrypted CCS which is not allowed.

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]

File: library/ssl_tls13_client.c
Comment (xkqian) [reply library/ssl_tls13_client.c:1987]: When ssl->early_data_status == MBEDTLS_SSL_EARLY_DATA_STATUS_ACCEPTED, seems we should not re-generate the early secrects. Should it be: 
```
if( ssl->early_data_status == MBEDTLS_SSL_EARLY_DATA_STATUS_NOT_SENT ||
        handshake->key_exchange_mode ==
            MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL )
```

File: library/ssl_tls13_client.c
Comment (xkqian) [reply library/ssl_tls13_client.c:2298]: This commnet leaves here for the future refining of the state machine changes, such as: we remove the co-ordinate function of client_certificate and the state **MBEDTLS_SSL_CLIENT_CERTIFICATE** is not a must, then we should change the `mbedtls_ssl_set_outbound_transform( ssl, ssl->handshake->transform_handshake );` to other place , see the comment: https://github.com/Mbed-TLS/mbedtls/pull/6499#discussion_r1066817923

File: library/ssl_tls13_client.c
Comment (xkqian) [reply library/ssl_tls13_client.c:1741]: ssl->handshake->ciphersuite_info will be updated to the chosen ciphersuite_info in line `1736: ssl->handshake->ciphersuite_info = ciphersuite_info;`, If we **didn't** add this check, both `ssl->session_negotiate->ciphersuite` and` ssl->handshake->ciphersuite_info->id` have been updated to the chosen ciphersuite here, and there are **nowhere** to store the  **ciphersuite of the ticket used for early data**.

File: library/ssl_tls13_client.c
Comment (xkqian) [reply library/ssl_tls13_client.c:1247]: RFC 8446  section D.4 said: 
-  **If not offering early data**, the client sends a dummy
      change_cipher_spec record (see the third paragraph of [Section 5](https://www.rfc-editor.org/rfc/rfc8446#section-5))
      immediately before its second flight.  This may either be before
      its second ClientHello or before its encrypted handshake flight.
--   **If offering early data**, the record is placed immediately after the
      first ClientHello.
Does that mean we should always send out the ccs, and the difference only exists in :where to send.

File: library/ssl_client.c
Comment (yuhaoth) [inline library/ssl_client.c:1002]:
```suggestion
#if defined(MBEDTLS_SSL_PROTO_TLS1_3)
    if( ssl->handshake->min_tls_version <=  MBEDTLS_SSL_VERSION_TLS1_3 &&
        MBEDTLS_SSL_VERSION_TLS1_3 <= ssl->tls_version )
    {
        /* TLS 1.3 is proposed */
        ret = mbedtls_ssl_tls13_finalize_write_client_hello( ssl );
    }
    else
#endif
    {
        mbedtls_ssl_handshake_set_state( ssl, MBEDTLS_SSL_SERVER_HELLO );
    }
```
A nitpick comment.

File: library/ssl_tls13_client.c
Comment (ronald-cron-arm) [reply library/ssl_tls13_client.c:1728]: You are right that's more subtle than I thought.

Unfortunately I do not think that's the current version is right either. In a resumption situation, we may end up with a ciphersuite different from the ciphersuite of the session that we resume: only the hash needs to match between the two ciphersuites if there is no early data involved. And in that case we need to update somewhere `ssl->session_negotiate->ciphersuite`.

I've played a bit with this and the only reasonable solution I came up with was to:
1. In ssl_tls13_generic/keys.c systematically use `ssl->handshake->ciphersuite_info` to retrieve the "current ciphersuite", never `ssl->session_negotiate->ciphersuite`.
2. Move the update of `ssl->session_negotiate->ciphersuite` to `ssl_tls13_postprocess_hrr()` and to `ssl_tls13_postprocess_server_hello()`, after having done the comparison with `ssl->handshake->ciphersuite_info->id` in the latter case.

File: library/ssl_tls13_client.c
Comment (ronald-cron-arm) [reply library/ssl_tls13_client.c:1990]: Yes you are right.

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]

File: library/ssl_tls13_client.c
Comment (xkqian) [reply library/ssl_tls13_client.c:1728]: How about the solution:
We only set ssl->session_negotiate->ciphersuite when not resumtion. 
For resume case, we only add item 2: 
"`Move the update of ssl->session_negotiate->ciphersuite to ssl_tls13_postprocess_hrr() and to ssl_tls13_postprocess_server_hello(), after having done the comparison with ssl->handshake->ciphersuite_info->id in the latter case.`" 
as:  "`ssl->session_negotiate->ciphersuite = ssl->handshake->ciphersuite_info->id`"

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]

File: library/ssl_tls13_client.c
Comment (xkqian) [reply library/ssl_tls13_client.c:1961]: Update done.
Anyway, I am confusing whether this description is for server only or for both server and client.
"**the server** MUST have accepted a PSK
         * cipher suite and selected the first key offered in the client’s
         * "pre_shared_key" extension. In addition, **it**..."

File: library/ssl_tls13_client.c
Comment (ronald-cron-arm) [inline library/ssl_tls13_client.c:1964]:
```suggestion
         * - The selected ALPN [RFC7301] protocol, if any
         *
         * We check here that when early data is involved the server
         * selected the cipher suite associated to the pre-shared key
         * as it must have.

File: library/ssl_tls13_client.c
Comment (ronald-cron-arm) [reply library/ssl_tls13_client.c]: You are right, as it is it could be ambiguous. I've suggested something to improve.

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
Thanks for all your work on this. Some final nitpicks, other than that this looks good to me. You need to rebase and go through the coding style change though.

**Review by ronald-cron-arm — APPROVED** [review]
LGTM. You need to rebase now though.

**Comment by xkqian** [comment]
> > And no tests for ciphersuite mismatch? If possible, please add it.
> 
> If we want to add the test case of ciphersuite mismatch, maybe we should add another option named "reco_ciphersuite" in ssl_client2. Shoud we place it in another PR, or any good suggestion?

Create a issue about early data test cases :https://github.com/Mbed-TLS/mbedtls/issues/7068 to follow up the ciphersuite mismatch case.

**Review by ronald-cron-arm — APPROVED** [review]
The rebase looks good to me. @yuhaoth please have another look. Then maybe one day we will get an useful feedback from the CIs concerning the merge jobs.

**Review by yuhaoth — APPROVED** [review]
LGTM

**Pull request merged**
